src/bases/base.js

```diff
@@ -0,0 +1,272 @@
+// @ts-check
+
+/**
+ * @typedef {import('./interface').BaseEncoder} BaseEncoder
+ * @typedef {import('./interface').BaseDecoder} BaseDecoder
+ * @typedef {import('./interface').BaseCodec} BaseCodec
+ */
+
+/**
+ * @template {string} T
+ * @typedef {import('./interface').Multibase<T>} Multibase
+ */
+/**
+ * @template {string} T
+ * @typedef {import('./interface').MultibaseEncoder<T>} MultibaseEncoder
+ */
+
+/**
+ * Class represents both BaseEncoder and MultibaseEncoder meaning it
+ * can be used to encode to multibase or base encode without multibase
+ * prefix.
+ * @class
+ * @template {string} Base
+ * @template {string} Prefix
+ * @implements {MultibaseEncoder<Prefix>}
+ * @implements {BaseEncoder}
+ */
+class Encoder {
+  /**
+   * @param {Base} name
+   * @param {Prefix} prefix
+   * @param {(bytes:Uint8Array) => string} baseEncode
+   */
+  constructor (name, prefix, baseEncode) {
+    this.name = name
+    this.prefix = prefix
+    this.baseEncode = baseEncode
+  }
+
+  /**
+   * @param {Uint8Array} bytes
+   * @returns {Multibase<Prefix>}
+   */
+  encode (bytes) {
+    if (bytes instanceof Uint8Array) {
+      return `${this.prefix}${this.baseEncode(bytes)}`
+    } else {
+      throw Error('Unknown type, must be binary type')
+    }
+  }
+}
+
+/**
```

Comment:
what's going on with these two blocks?

Reply:
are they setting up "variables" for the block below by hoisting them out of `interface`? we don't get automatic linkage by having common names across the project so they have to be explicitly linked?

Reply:
oh, and how are you verifying this? do you have some kind of linting tool that tells you when you're missing symbols and other entries that need to be pulled out?

Reply:
`// @ts-check` on the first line tells vscode to use typescript to do this. If project had `tsconfig.json` that would not be necessary and it would that to.